inc/cappuccino/TSet.hpp

```diff
@@ -0,0 +1,520 @@
+    std::chrono::milliseconds m_ttl;
+};
+} // namespace cappuccino
+
+namespace cappuccino
+{
+
+template <typename KeyType, Sync SyncType>
+TSet<KeyType, SyncType>::TSet(std::chrono::milliseconds ttl, float max_load_factor) 
+{
+    m_keyed_elements.max_load_factor(max_load_factor);
+    m_ttl = ttl;
+}
+
+template <typename KeyType, Sync SyncType>
+auto TSet<KeyType, SyncType>::Insert(
+        const KeyType& key,
+        std::chrono::milliseconds ttl,
+        Allow allow) -> bool
+{
+    std::lock_guard guard { this->m_lock };
+    const auto now = std::chrono::steady_clock::now();
+    const auto expire_time = now + ttl;
+
+    this->doPrune(now);
+
+    return this->doInsertUpdate(key, expire_time, allow);
+}
+
+template <typename KeyType, Sync SyncType>
+auto TSet<KeyType, SyncType>::Insert(
+        const KeyType& key,
+        Allow allow) -> bool
+{
+    return Insert(key, m_ttl, allow);
+}
+
+template <typename KeyType, Sync SyncType>
+template <typename RangeType>
+auto TSet<KeyType, SyncType>::InsertRange(
+        RangeType&& key_range,
+        std::chrono::milliseconds ttl,
+        Allow allow) -> size_t
+{
+    size_t inserted { 0 };
+    std::lock_guard guard { this->m_lock };
+    const auto now = std::chrono::steady_clock::now();
+    const auto expire_time = now + ttl;
+
+    this->doPrune(now);
+
+    for (const auto& key  : key_range) {
+        if (this->doInsertUpdate(key, expire_time, allow))
+        {
+            ++inserted;
+        }
+    }
+
+    return inserted;
+}
+
+template <typename KeyType, Sync SyncType>
+template <typename RangeType>
+auto TSet<KeyType, SyncType>::InsertRange(
+        RangeType&& key_range,
+        Allow allow) -> size_t
+{
+    return InsertRange(std::move(key_range), m_ttl, allow);
+}
+
+template <typename KeyType, Sync SyncType>
+auto TSet<KeyType, SyncType>::Delete(
+    const KeyType& key) -> bool
+{
+    std::lock_guard guard { m_lock };
+    const auto now = std::chrono::steady_clock::now();
+
+    doPrune(now);
+
+    const auto keyed_position = m_keyed_elements.find(key);
+    if (keyed_position != m_keyed_elements.end()) {
+        doDelete(keyed_position);
+        return true;
+    } else {
+        return false;
+    }
+};
+
+template <typename KeyType, Sync SyncType>
+template <template <class...> typename RangeType>
+auto TSet<KeyType, SyncType>::DeleteRange(
+    const RangeType<KeyType>& key_range) -> size_t
+{
+    size_t deleted { 0 };
+    std::lock_guard guard { m_lock };
+    const auto now = std::chrono::steady_clock::now();
+
+    doPrune(now);
+
+    for (const auto& key : key_range) {
+        const auto keyed_position = m_keyed_elements.find(key);
+        if (keyed_position != m_keyed_elements.end()) {
+            doDelete(keyed_position);
+            ++deleted;
+        }
+    }
+
+    return deleted;
+};
+
+template <typename KeyType, Sync SyncType>
+auto TSet<KeyType, SyncType>::Find(
+    const KeyType& key) -> bool
```

Comment:
I think the API for find on the set should mimic https://en.cppreference.com/w/cpp/container/set/find and return the item in the cache just like the map+cache types do.

Reply:
On second thought bool might be better here, the STL returns an iterator which points to the item in the set, bool seems wholly appropriate since the user is passing the key in by const& anyways.